### PR TITLE
[hotfix] Fix Redis Sink to fail at opening if Redis is not initialized.

### PR DESCRIPTION
**Pull request by kl0u**
Before if Redis was not initialized, the job would start and would wait until the first element arrives in order to realize that Redis is not running. This fixes that by checking the connection to redis during the opening phase.

### CODE REVIEW

File: flink-streaming-connectors/flink-connector-redis/src/main/java/org/apache/flink/streaming/connectors/redis/RedisSink.java [hunk 1]
```diff
@@ -166,7 +166,13 @@ public void invoke(IN input) throws Exception {
      */
 	@Override
 	public void open(Configuration parameters) throws Exception {
-		this.redisCommandsContainer = RedisCommandsContainerBuilder.build(this.flinkJedisConfigBase);
+		try {
+			this.redisCommandsContainer = RedisCommandsContainerBuilder.build(this.flinkJedisConfigBase);
+			this.redisCommandsContainer.open();
+		} catch (Exception e) {
+			LOG.error("Redis has not been properly initialized.");
```
Comment (tillrohrmann) [inline flink-streaming-connectors/flink-connector-redis/src/main/java/org/apache/flink/streaming/connectors/redis/RedisSink.java]: The exception should be logged here as well.

File: flink-streaming-connectors/flink-connector-redis/src/main/java/org/apache/flink/streaming/connectors/redis/common/container/RedisClusterContainer.java [hunk 2]
```diff
@@ -47,6 +47,11 @@ public RedisClusterContainer(JedisCluster jedisCluster) {
 	}
 
 	@Override
+	public void open() throws Exception {
+		jedisCluster.echo("Test");
```
Comment (tillrohrmann) [inline flink-streaming-connectors/flink-connector-redis/src/main/java/org/apache/flink/streaming/connectors/redis/common/container/RedisClusterContainer.java]: testing artifact?

File: flink-streaming-connectors/flink-connector-redis/src/main/java/org/apache/flink/streaming/connectors/redis/common/container/RedisContainer.java [hunk 3]
```diff
@@ -68,12 +68,12 @@ public RedisContainer(final JedisSentinelPool sentinelPool) {
...
-		}
-		if (this.jedisSentinelPool != null) {
-			this.jedisSentinelPool.close();
-		}
+		getInstance().close();
+	}
+
+	@Override
+	public void open() throws Exception {
+		getInstance().echo("Test");
```
Comment (tillrohrmann) [inline flink-streaming-connectors/flink-connector-redis/src/main/java/org/apache/flink/streaming/connectors/redis/common/container/RedisContainer.java]: `echo("Test")`?

Comment (kl0u) [inline flink-streaming-connectors/flink-connector-redis/src/main/java/org/apache/flink/streaming/connectors/redis/common/container/RedisContainer.java]: Yes. It tries to open a connection and just echo the message "Test". 
This is to see if there is a way to communicate with the cluster or server (depending on the setup).

File: flink-streaming-connectors/flink-connector-redis/src/main/java/org/apache/flink/streaming/connectors/redis/common/container/RedisContainer.java [hunk 4]
```diff
@@ -68,12 +68,12 @@ public RedisContainer(final JedisSentinelPool sentinelPool) {
 	 */
 	@Override
 	public void close() throws IOException {
-		if (this.jedisPool != null) {
-			this.jedisPool.close();
-		}
-		if (this.jedisSentinelPool != null) {
-			this.jedisSentinelPool.close();
-		}
+		getInstance().close();
```
Comment (tillrohrmann) [inline flink-streaming-connectors/flink-connector-redis/src/main/java/org/apache/flink/streaming/connectors/redis/common/container/RedisContainer.java]: Is this equivalent to the old `close` method definition?

File: flink-streaming-connectors/flink-connector-redis/src/main/java/org/apache/flink/streaming/connectors/redis/common/container/RedisContainer.java [hunk 5]
```diff
@@ -68,12 +68,12 @@ public RedisContainer(final JedisSentinelPool sentinelPool) {
 	 */
```
Comment (tillrohrmann) [inline flink-streaming-connectors/flink-connector-redis/src/main/java/org/apache/flink/streaming/connectors/redis/common/container/RedisContainer.java]: The `RedisContainer` class contains fields which are not serializable. The `jedisPool` and the `jedisSentinelPool`.

File: flink-streaming-connectors/flink-connector-redis/src/main/java/org/apache/flink/streaming/connectors/redis/common/container/RedisClusterContainer.java [hunk 6]
```diff
@@ -47,6 +47,11 @@ public RedisClusterContainer(JedisCluster jedisCluster) {
 	}
```
Comment (tillrohrmann) [inline flink-streaming-connectors/flink-connector-redis/src/main/java/org/apache/flink/streaming/connectors/redis/common/container/RedisClusterContainer.java]: `jedisCluster` is not serializable

File: flink-streaming-connectors/flink-connector-redis/src/main/java/org/apache/flink/streaming/connectors/redis/common/container/RedisContainer.java (hunk 3 above)
Comment (tillrohrmann) [inline flink-streaming-connectors/flink-connector-redis/src/main/java/org/apache/flink/streaming/connectors/redis/common/container/RedisContainer.java]: Is this the only way to test a connection? If so, then it should be documented properly that this is used to test the connection.

**Comment by tillrohrmann** [comment]
Thanks for your contribution @kl0u. I had some inline comments.

File: flink-streaming-connectors/flink-connector-redis/src/main/java/org/apache/flink/streaming/connectors/redis/common/container/RedisContainer.java (hunk 4 above)
Comment (kl0u) [inline flink-streaming-connectors/flink-connector-redis/src/main/java/org/apache/flink/streaming/connectors/redis/common/container/RedisContainer.java]: Yes because from the constructors, only one of the `jedisSentinelPool` or `jedisPool` can be non-null value and that is what `getInstance()` returns.

**Comment by kl0u** [comment]
Hi @tillrohrmann !Thanks for the comments, I addressed the ones that should be addressed.

File: flink-streaming-connectors/flink-connector-redis/src/test/java/org/apache/flink/streaming/connectors/redis/RedisSinkTest.java [hunk 7]
```diff
@@ -41,6 +50,73 @@ public void shouldThrowNullPointerExceptionIfConfigurationIsNull(){
...
+
+		try {
+			redisSink.open(new Configuration());
+		} catch (Throwable e) {
+
+			// search for nested ConnectionExceptions
+			// because this is the expected behavior
+
+			int depth = 0;
+			while (!(e instanceof JedisConnectionException)) {
```
Comment (tillrohrmann) [inline flink-streaming-connectors/flink-connector-redis/src/test/java/org/apache/flink/streaming/connectors/redis/RedisSinkTest.java]: This condition will always be true/false if you don't reset `e = cause` in the while loop body.

**Comment by tillrohrmann** [comment]
Changes look good @kl0u. Good work :-) There is only a minor issue with one of the test cases and its exception handling.

**Comment by kl0u** [comment]
Thanks a lot @tillrohrmann ! 
I integrated your comment.

File: flink-streaming-connectors/flink-connector-redis/src/test/java/org/apache/flink/streaming/connectors/redis/RedisSinkTest.java [hunk 8]
```diff
@@ -103,15 +105,17 @@ private void testDownBehavior(FlinkJedisConfigBase config) throws Exception {
...
 
 			int depth = 0;
 			while (!(e instanceof JedisConnectionException)) {
-				Throwable cause = e.getCause();
-				if (cause == null || depth++ == 20) {
-					throw e;
+				e = e.getCause();
+				if (e == null || depth++ == 20) {
+					init.printStackTrace();
+					fail("Test Failed: " + init.getMessage());
```
Comment (tillrohrmann) [inline flink-streaming-connectors/flink-connector-redis/src/test/java/org/apache/flink/streaming/connectors/redis/RedisSinkTest.java]: Why don't we simply throw the initial exception here?

**Comment by kl0u** [comment]
Done! Thanks @tillrohrmann

File: flink-streaming-connectors/flink-connector-redis/src/test/java/org/apache/flink/streaming/connectors/redis/RedisSinkTest.java [hunk 9]
```diff
@@ -105,18 +105,13 @@ private void testDownBehavior(FlinkJedisConfigBase config) throws Exception {
...
-			while (!(e instanceof JedisConnectionException)) {
-				e = e.getCause();
-				if (e == null || depth++ == 20) {
-					init.printStackTrace();
-					fail("Test Failed: " + init.getMessage());
-				}
+			
+			if (!(e instanceof JedisConnectionException)) {
+				e.printStackTrace();
+				fail("Test Failed: " + e.getMessage());
```
Comment (tillrohrmann) [inline flink-streaming-connectors/flink-connector-redis/src/test/java/org/apache/flink/streaming/connectors/redis/RedisSinkTest.java]: Why did you change the `while` loop into a `if` condition. Did change something that the `JedisConnectionException` will always be the top level exception?

Comment (tillrohrmann) [inline flink-streaming-connectors/flink-connector-redis/src/test/java/org/apache/flink/streaming/connectors/redis/RedisSinkTest.java]: I think it is better to let the exception bubble up instead of printing the stack trace and calling `fail`.

Comment (kl0u) [inline flink-streaming-connectors/flink-connector-redis/src/test/java/org/apache/flink/streaming/connectors/redis/RedisSinkTest.java]: I also thought so that is why I had it like this but I thought your previous comment was to remove it.
Probably i misunderstood. Actually the exception was in the first layer since the beginning.
I can change it back ;)

**Comment by kl0u** [comment]
I think this is ok now.
